storage/execout/file.go

```diff
@@ -2,213 +2,332 @@ package execout
-			return value.Payload, true
+	item, err := streamproto.ReadNextItem(fr.reader)
+	if err != nil {
+		if err == io.EOF {
+			fr.complete = true
+			return nil, io.EOF
 		}
+		return nil, err
 	}
-
-	return nil, false
+	fr.lastReadItem = item
+	return item, nil
 }
 
-var MaxExecoutSegmentSize = int(8589934592)
-var ErrSegmentSizeExceeded = connect.NewError(connect.CodeInvalidArgument, fmt.Errorf("execution output segment size exceeded %d bytes: substreams cannot process this segment", MaxExecoutSegmentSize))
-
-func (c *File) Load(ctx context.Context) error {
-	c.Lock()
-	defer c.Unlock()
-	if c.loaded {
-		return nil
+func rewriteAsOrdered(ctx context.Context, r io.Reader, readBytes []byte, store dstore.Store, moduleName, filename string, rng *block.Range, logger *zap.Logger) error {
+	bytes, err := io.ReadAll(r)
+	if err != nil {
+		return fmt.Errorf("reading store file %s: %w", filename, err)
 	}
+	bytes = append(readBytes, bytes...)
 
-	filename := computeDBinFilename(c.Range.StartBlock, c.Range.ExclusiveEndBlock)
-	c.logger.Debug("loading execout file", zap.String("file_name", filename), zap.Object("block_range", c.Range))
-
-	err := derr.RetryContext(ctx, 5, func(ctx context.Context) error {
-		objectReader, err := c.store.OpenObject(ctx, filename)
-		if err == dstore.ErrNotFound {
-			return derr.NewFatalError(err)
-		}
-
-		if err != nil {
-			return fmt.Errorf("loading block reader %s: %w", filename, err)
-		}
-		defer objectReader.Close()
+	o := &pboutput.Array{}
+	if err := o.UnmarshalVTUnsafe(bytes); err != nil {
+		return fmt.Errorf("unmarshalling data: %w", err)
+	}
+	items := o.Items
 
-		// Limit reading to MaxExecoutSegmentSize to prevent excessive memory usage
-		limitedReader := io.LimitReader(objectReader, int64(MaxExecoutSegmentSize+1))
-		bytes, err := io.ReadAll(limitedReader)
-		if err != nil {
-			return fmt.Errorf("reading store file %s: %w", filename, err)
-		}
-		if len(bytes) == MaxExecoutSegmentSize+1 {
-			return derr.NewFatalError(fmt.Errorf("%w: file %s on module %s", ErrSegmentSizeExceeded, filename, c.ModuleName))
-		}
+	sort.Slice(items, func(i, j int) bool {
+		return items[i].BlockNum < items[j].BlockNum
+	})
 
-		outputData := &pboutput.Map{}
-		if err = outputData.UnmarshalFast(bytes); err != nil {
-			return fmt.Errorf("unmarshalling file %s: %w", filename, err)
+	store.SetOverwrite(true)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel() // closes the WriteAsYouGo
+	newFile := NewFileWriter(ctx, store, logger, rng, moduleName)
+
+	for _, item := range items {
+		if err := newFile.SetItem(&pbsubstreams.Clock{
+			Id:        item.BlockId,
+			Number:    uint64(item.BlockNum),
+			Timestamp: item.Timestamp,
+		}, item.Payload); err != nil {
+			return err
 		}
-
-		c.Kv = outputData.Kv
-
-		c.logger.Debug("outputs data loaded", zap.Int("output_count", len(c.Kv)), zap.Stringer("block_range", c.Range))
-		return nil
-	})
-	if err == nil {
-		c.loaded = true
 	}
-	return err
+	return newFile.Close()
```

Comment:
There is an exit path in the for/loop above, a defer should be used to close the file.